packages/drivers/odsp-driver/src/fetchSnapshot.ts

```diff
@@ -239,14 +240,40 @@ async function redeemSharingLink(
 				const encodedShareUrl = getEncodedShareUrl(
 					odspResolvedUrl.shareLinkInfo?.sharingLinkToRedeem,
 				);
-				const redeemUrl = `${odspResolvedUrl.siteUrl}/_api/v2.0/shares/${encodedShareUrl}`;
-				const { url, headers } = getUrlAndHeadersWithAuth(
-					redeemUrl,
-					storageToken,
-					forceAccessTokenViaAuthorizationHeader,
-				);
-				headers.prefer = "redeemSharingLink";
-				return fetchAndParseAsJSONHelper(url, { headers });
+
+				let redeemUrl: string | undefined;
+				async function callSharesAPI(baseUrl: string): Promise<void> {
+					redeemUrl = `${baseUrl}/_api/v2.0/shares/${encodedShareUrl}`;
+					const { url, headers } = getUrlAndHeadersWithAuth(
+						redeemUrl,
+						storageToken,
+						forceAccessTokenViaAuthorizationHeader,
+					);
+					headers.prefer = "redeemSharingLink";
+					await fetchAndParseAsJSONHelper(url, { headers });
+				}
+
+				const disableUsingTenantDomain = loggerToMonitoringContext(
+					logger,
+				).config.getBoolean("Fluid.Driver.Odsp.DisableUsingTenantDomainForSharesApi");
+				// There is an issue where if we use the siteUrl in /shares, then the allowed length of url is just a few hundred characters
+				// and we fail to do the redeem. But if we use the tenant domain in the url, then the allowed length becomes 2048. So, first
+				// construct the url for /shares using tenant domain but to be on safer side, fallback to what we were using before.
+				if (!disableUsingTenantDomain) {
+					try {
+						await callSharesAPI(new URL(odspResolvedUrl.siteUrl).origin);
+						return;
+					} catch (error) {
+						logger.sendTelemetryEvent(
+							{
+								eventName: "ShareLinkRedeemFailedWithDomain",
+								length: redeemUrl?.length,
+							},
+							error,
+						);
+					}
+				}
+				await callSharesAPI(odspResolvedUrl.siteUrl);
```

Comment:
In case user has no permissions, we will do 2 calls. Is that Ok?

Reply:
The other issue (which might be fine) - if first call fails because client was offline, we may do a second call when we are not offline.
Ideally the fork about which API to use should be much higher, above the place where we have retry loop (I assume we have retry loop for that flow).

Reply:
I think it is ok. It is a fallback case.